WIP: initial attempt to run 'esm' against tc39/test262

dnalborczyk opened the pull request:
**work in progress**

_notes and questions coming soon_

just want to see some runs first :yum:

**update:**

alrighty, cool. looking pretty good so far! 

_btw, the folder structure etc. where test262 is being cloned into and modified to etc. should probably change. also the way the tests are run (right now as part of the whole test suite). we can separate it out and make it a separate run, or a combination thereof..._

couple things I'm noticing which are still to do:
- [x] skiplist not working on Windows
- [x] activate other folders, e.g. `import`, `export`
- [x] fix globbing (missed subfolders)
- [x] make test262 pass on each platform first (use skiplist), then go backwards and unblock
- [x] add --harmony flag for `private class property` on supported platforms only (node v10+ probably)
  ~- [ ] use test262 frontmatter `features` class-fields-private~
- [ ] should we recognize the frontmatter `module`-flag?
  - [ ] investigate: there might be other tests in other folders with that flag
- [x] investigate `node v6.2` failures: probably unsupported features [yep, works on v6.14.3 LTS latest]
- [x] add with esm options: { cjs:false }
- [x] investigate: the `chakracore` run
- [ ] implement flag for ignoring (all or portions) of the skiplist
- [x] fetch `dynamic import` test [branch from PR](https://github.com/tc39/test262/pull/1588)
- [x] run against `--experimental-modules`

_I'll add more later ..._

jdalton commented on test/tests.js:
```diff
@@ -18,6 +18,7 @@ import "./main-hook-tests.mjs"
 import "./repl-hook-tests.mjs"
 import "./scenario-tests.mjs"
+import "./_external/test262/index.js"
 
```
We call the external folder in `src` the `vendor` folder. We could call it that in test too.

dnalborczyk replied:
will do.

jdalton commented on test/_external/test262/blacklist:
```diff
@@ -0,0 +1,101 @@
+# export * from ns
```
Could you rename this list to `skiplist`

dnalborczyk replied:
done.

dnalborczyk commented:
just for tracking purposes, this one just broke in recent commits: https://github.com/tc39/test262/blob/master/test/language/module-code/parse-err-yield.js

maybe one of the `acorn` patches? want me to add it to the `skiplist` for the interim?

jdalton commented:
> maybe one of the acorn patches? want me to add it to the skiplist for the interim? d262630 we can unlock it later again ...

Patched https://github.com/standard-things/esm/commit/3124369032d03b70f700e17d6981192ae6b184ae.

dnalborczyk commented:
@jdalton not sure if `travis` and `appveyor` acknowledges `git push --force`, but have you considered:

https://www.appveyor.com/docs/build-configuration/#rolling-builds
https://blog.travis-ci.com/2017-03-22-introducing-auto-cancellation

both are slightly different in semantics, but have a similar goal.

dnalborczyk commented:
**update:**

I have `good news` and `bad news`.

`good news` is that we have `more tests` now. `bad news` is also `more failures`.

I missed the subfolders in the initial globbing. I thought the tests where only in the roots. :confused:  I also included the `import` and `export` folders.

anyhow, initial test262 tests from `module-code` only (w/o subfolders):
```
218 passing
34 pending
```

with `import` and `export` folders:
```
225 passing
34 pending
```

fixed globbing (including subfolders, `module-code` only)
```
249 passing
46 pending
```

2 of the new failures are `export * from ns` failures (6 overall now). some of the failures are platform (V8) specific (e.g. private class property). some of the failures are unrelated to the test case like you mentioned. one fix might fix several test cases. I can't also guarantee [yet] that the code generating the tests is correct.

dnalborczyk commented:
cool, awesome!! I also rebased against master and removed the skiplist. your recent `bug fixes` (cycle, self import) etc. made some of them already pass. I'll update the skiplist to reflect that.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -135,3 +135,6 @@ test/language/module-code/namespace/internals/own-property-keys-sort.js
 test/language/module-code/namespace/internals/set.js
+
+# NEEDS INVESTIGATION! -> only fails on node v6.14.3
+test/language/module-code/namespace/internals/object-keys-binding-uninit.js
```
This is just a V8 tweak. In older Node they didn't error. In newer ones they do. We support erroring whenever Node does so this can be gated to whatever Node it's supported in. Related to https://bugs.chromium.org/p/v8/issues/detail?id=7780.

dnalborczyk replied:
ah, ok. it seems to be on the v6 branch only I believe. I'll implement a condition for _allowed_ versions or something like that (e.g. node6, harmony, etc.)

jdalton commented on .travis.yml:
```diff
@@ -7,7 +7,7 @@ os:
   - 10
   - 8
-  - 6.2.0
+  - 6
```
we'll want to keep that low 6.2.0 is the lowest I can run the full suite at without random seg faults (due to issues in earlier 6.0 releases). It's fine as Node official support is only whatever the latest 6 is but for nice coverage I like to test as far back as possible.

dnalborczyk replied:
yeah, that was just temporary. did you see my question regarding that toppic --> https://github.com/standard-things/esm/pull/550#issuecomment-410850973

I'll extend the skiplist to be conditional. not: all or nothing

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -99,3 +105,33 @@ test/language/module-code/instn-star-props-nrml.js
+test/language/module-code/namespace/internals/get-str-update.js
+
+# NEEDS INVESTIGATION!
+test/language/module-code/namespace/internals/own-property-keys-sort.js
```
We assign namespace keys in sort order so I'd be interested in that fail. 👆

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -99,3 +105,33 @@ test/language/module-code/instn-star-props-nrml.js
+test/language/module-code/namespace/internals/own-property-keys-sort.js
+
+# NEEDS INVESTIGATION!
+test/language/module-code/namespace/internals/set.js
```
We throw here too but be careful as by default `esm` enables CJS interop with mutable namespace objects (which don't throw). This is configurable though `cjs.mutableNamespace`.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -99,3 +105,33 @@ test/language/module-code/instn-star-props-nrml.js
 
 # NEEDS INVESTIGATION!
 test/language/module-code/parse-err-reference.js
+
```
Many of these fails look to be related to having `cjs.mutableNamespace` enabled 👇

dnalborczyk replied:
ah. ok. that's good tho! what's the plan forward to make them pass? is that with `mjs` disabled as well? I would assume `mjs` and `use module` work the same (or similar)?

jdalton replied:
to avoid `.mjs` you could just include ` .esmrc` or `.esmrc.json` in the test files folder or parent folder with specific options disabled.

dnalborczyk commented:
_latest [master](3124369032d03b70f700e17d6981192ae6b184ae): (minus chakracore)_

```
255 passing
40 pending
```

dnalborczyk commented:
added `dynamic import` tests from PR branch:

unfortunately it didn't add many additional tests [hopefully more are coming].

status: 2 failing, 1 passing.

overall:
256 passing
42 pending

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -1,3 +1,9 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/dynamic-import/eval-self-once.js
+
```
They use a test harness method called `fnGlobalObject` 👆

dnalborczyk replied:
I believe `fnGlobalObject` should be injected. I think the test failure was saying that the module was being loaded several times. let me check.

jdalton replied:
In my local test of `eval-self-once` (replacing the global helper with `Function("return this")()`, turning the `assert` into a `console.log` and commenting out the `.then($DONE,$DONE)` I get

```
1 1 'global property was defined and incremented only once'
```

jdalton replied:
We have tests for eval count ourselves too which is why I'm keying in on this skip as something odd.

dnalborczyk replied:
aaah. sorry. I think I saw something with `$DONE` being `undefined`

dnalborczyk replied:
I'll have another look.

dnalborczyk replied:
I must have missed that: 
https://github.com/tc39/test262/blob/master/harness/doneprintHandle.js

thanks for pointing it out!

I didn't look into it further at that time as I wanted the test run to be green first - and then investigate. so there's a chance that some other tests are failing because of that or similar issues as well.

jdalton replied:
That's cool. (We have tests covering this so it's less critical).

dnalborczyk replied:
I looked into the failure. the reason was that the requested-once-`import` and requested-once-`dynamic import` tests are using the same global variable `global.test262`. the `import` tests would set it (and don't remove it afterwards), which makes the `dynamic import` tests fail.

I'll remove the `global.test262` for now in every it(...). not sure if every test is supposed to run in a fresh and clean context.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,125 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/eval-self-abrupt.js
+
```
👆 is using `throw new Test262Error`

dnalborczyk replied:
here is the result:
```shell
1) test262 module tests
       Abrupt completion from module evaluation is reported (eval-self-abrupt.js) :

      AssertionError [ERR_ASSERTION]: Input A expected to strictly equal input B:
+ expected - actual

- 'ReferenceError'
+ 'Test262Error'
      + expected - actual

      -ReferenceError
      +Test262Error
```

jdalton replied:
It's a reference error if `Test262Error ` doesn't exist.
<img width="375" src="https://user-images.githubusercontent.com/4303/43810508-ecb656ce-9a6c-11e8-93c3-28936d3b9077.png">

dnalborczyk replied:
uh, my bad. 😞
I thought the Test262 exceptions are always a string. guess not. fixing ...
thanks!

dnalborczyk replied:
verified ✅ 

this was some regression on my side. Test262Error was being injected (just not always, depending on the flag). I'll fix properly later, if needed.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,125 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/instn-star-as-props-dflt-skip.js
+
```
👆 `export * as named from` is in `test/language/module-code/instn-star-as-props-dflt-skip.js`

dnalborczyk replied:
thanks, updated!

dnalborczyk replied:
just noticed, `export * from` vs. `export * as ns` ... fixed.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,125 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/instn-star-ambiguous.js
+
```
👆 This is a bug (patching)

jdalton replied:
Patched https://github.com/standard-things/esm/commit/599b0df265c38445051a4d4923cad3157d70824a

dnalborczyk commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,125 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/instn-iee-bndng-var.js
+
```
checking ....

dnalborczyk replied:
not sure if you changed something, or my rebase was bad:

throws now:
```
file:///Users/daniel/dev/github/esm_fork/test/_external/test262/temp/test/language/module-code/instn-iee-bndng-var.js:136
  y,

ReferenceError: y is not defined
    at file:///Users/daniel/dev/github/esm_fork/test/_external/test262/temp/test/language/module-code/instn-iee-bndng-var.js:136:9
    at Generator.next (<anonymous>)
    1) Imported binding reflects state of indirectly-exported `var` binding
 (instn-iee-bndng-var.js)
```

jdalton replied:
Ah yep, it turns out those 3 tests won't be fully fixed until I finish the instantiation phase work.

dnalborczyk commented:
alrighty, time for bed! 😃

we're definitely gettin' there ... 

```shell
266 passing
  32 pending
```

btw, I still haven't tried applying the suggested `esm options` yet ...

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,95 @@
+# NEEDS INVESTIGATION!
+test/language/module-code/eval-gtbndng-indirect-update-dflt.js
+
```
👆`esm` :bug:

jdalton replied:
Patched https://github.com/standard-things/esm/commit/e9d8fa58b2184ac13f84bda991d5ce3064ea7d0c.

jdalton commented on test/_external/test262/skiplist:
```diff
@@ -0,0 +1,95 @@
+
+# NEEDS INVESTIGATION!
+test/language/module-code/instn-iee-iee-cycle.js
+
```
👆`esm` :bug:

jdalton commented:
The remaining fails fall into a few buckets:

  * the instantiate phase work (8 fails)
  * bug test/language/module-code/instn-iee-iee-cycle.js
  * bug test/language/module-code/eval-gtbndng-indirect-update-dflt.js
  * setting `cjs.mutableNamespace` to `false` will fix (11 fails)
  * skipping export * as ns from and private fields tests (8 fails)

So that's 27 of the fails accounted for.

dnalborczyk commented:
> the instantiate phase work (8 fails)
> bug test/language/module-code/instn-iee-iee-cycle.js
> bug test/language/module-code/eval-gtbndng-indirect-update-dflt.js

I added those to the comment section of the skiplist as `CONFIRMED BUGS`

> setting cjs.mutableNamespace to false will fix (11 fails)

I'll give that a try

> skipping export * as ns from and private fields tests (8 fails)

marked as `UNSUPPORTED  (export * as ns from "mod")` for now

> So that's 27 of the fails accounted for (I could have miscounted a couple too).

problem is that the test262 file names look like each other 😃I mixed several up as well.

dnalborczyk commented:
I re-wrote the test run slightly, in order to simplify some things, after I have a better understanding how test262 is meant to be run, and also how the `cjs` and `mode` options are meant to work. I really have to go back to the documentation branch and keep going on that one. I have myself so many questions about how certain knobs [options] are interfering with the outcome.

anyhow, instead of injecting the harness code "with a hammer", after [reading this](https://github.com/tc39/test262/blob/master/INTERPRETING.md) I opted to "install" the harness on the `global` object with `runInThisContext` and an `esm wrapper` to apply `esm options`.

for some reason it made 2 tests fail:  instn-named-star-cycle.js + instn-named-star-cycle.js

I was wondering if my thinking is correct and this should be allowed?

it's something like this (simplified):

```
// test262-test.js
assert(this, undefined)
```


```javascript
// wrapper.js
require = require("esm")(module, { cjs: false, mode: "all" })
module.exports = require("./test.js")
```

```javascript
// test.js
export default function loadTest(url) {
  return import(url)
}
```

```javascript
it(...`, function () {
  return wrapper("some-path/test262-test.js")
    .catch((e) => {
```

jdalton commented on test/_external/test262/wrapper.js:
```diff
@@ -0,0 +1,2 @@
+require = require("esm")(module, { cjs: false, mode: "all" })
+module.exports = require("./test.js")
```
Can't use `esm` or that's the version of `esm` included with `ava`.
You need to usr `../../../index.js` or so.

jdalton commented:
> I forgot: `cjs:false` (which includes `cjs.mutableNamespace` I suppose, has not effect on the 11 test cases. they are failing nonetheless.

Ah, it's because of a small issue (bug). I'll resolve it soon.

jdalton commented:
Patched https://github.com/standard-things/esm/commit/0226cf453760818f960270922413a37fe279eda4.

dnalborczyk commented:
that seems to have fixed it. verifying ...

it [might have] created another issue with the `global` object. in the failing test cases case the `global.print` function (which I formerly "attached") is possibly missing. but the other `runInThisContext` test262 methods seem to be accessible (on the global).

dnalborczyk commented:
alrighty, I added some [cheap] flags for [including] `--harmony` and [excluding] `node-chakracore`.

status:
**node v6 - v10 (all platforms)**
```shell
257 passing
41 pending
```

**node v10 harmony (linux/mac)**
```shell
266 passing
32 pending
```

**node chakracore (windows)**
```shell
238 passing
60 pending
```

jdalton commented:
Thanks @dnalborczyk!

Did you resolve the `global.print` thing?

dnalborczyk commented:
`global.print` seems to be gone (inaccessible) in the wrapped module.

I'll digg deeper in a bit and give more feedback.

dnalborczyk commented:
@jdalton I have no clue what's going on. I don't even know how to describe it.

I attached print with `Reflect.set(global, "print", print)`, which seems to work in the same scope, but is gone in another. when I attach it with `global.print = print` it works.

in short:
```javascript
Reflect.set(global, "print", print)
Reflect.get(global, "print") // works as well

but once entering the next [function] scope, it's gone.
```

I have no clue if I'm missing something, or if that is a `bug` or intentional, or unintentional behavior of `Mocha`, `V8`, `node.js`, `esm` ... no clue. I was under the assumption that `Reflect.set` does the same job as setting a property directly.

dnalborczyk commented:
@jdalton   we got more `dynamic import` test cases!!! 😄 unfortunately also more failures, but that could easily be the test runner itself, since we had only a handful async test262 tests.

seems like **rwaldron** generated more tests in the PR: https://github.com/tc39/test262/pull/1588/commits/269908b039a11e806b5e9fe796424b7b7687f2d2

I'll look into it soon...

dnalborczyk commented:
**update:** turns out that `some tests` are sharing `fixtures` which each other, so if one module was being loaded, and the module scope modified _(turned out to be a bug)_, it would remain modified for remaining tests, which would make them fail as well, even though if run independently on their own, they would pass. I'll fix the runner to run independently of module caches [and whatever else might be in the way] with `node` and the `require` flag. that way it also doesn't pollute the `global scope` with the harness. that'll bring the failures down, and the numbers should be correct.

jdalton commented:
Wow, awesome work digging into it!

I've fixed a few more of the namespace related tests, working on another namespace one before I publish a release.

dnalborczyk commented:
**test262 | v3.0.75 status**

_excluding_ `dynamic import` tests  [that's why we're 3 tests short here compared to last time]:

anyhow, 7 more tests are passing!!

**node v6 - v10 (all platforms)**
```shell
261 passing
34 pending
```

**node v10 harmony (linux/macos)**
```shell
270 passing
25 pending
```

**node chakracore (windows)**
[although we need a run w/o the skiplist and see if anything _more_ is passing now]
```shell
238 passing
57 pending
```

<br />

_including_ `dynamic import` tests:

**node v10.8.0 harmony (macos)** [local run]
```shell
408 passing
25 pending
29 failing
```

I'll have to look into those and add them to the skiplist. 11 of the failing tests include a test with `import("")` (empty string) and they behave correct I believe.

dnalborczyk commented:
just found some time to run against `--experimental-modules`, and it's fairing pretty good:

**`node --experimental-modules --harmony`**  v10.8.0
```shell
280 passing
  8 pending  # excludes export * as ns from 'mod' tests
  7 failing
```

then I did a run against `esm with mjs, [mode:strict]`

**node -r esm --harmony**  v10.8.0

```shell
251 passing
  8 pending  # excludes export * as ns from 'mod' tests
  36 failing
```

it seems 2 more are failing than the test run with `js` extensions `(cjs: false, mode: "all"}`, and maybe even different cases, but I haven't compared any yet. I'll do a run with the `js`-skiplist as `mjs`-skiplist, and compare. if we have 2 more, then it's 2 more, if it's more than that, then we have different failures.

jdalton commented:
Big Big thanks @dnalborczyk!

I'm going to merge this in and then perform some cleanup on it.

dnalborczyk commented:
you're welcome! thank you John! @jdalton 

yeah, it probably needs some cleanup.

sorry for being a little inactive lately, just started a new job, and I'm knee deep occupied with the on-boarding process and diving into their codebase. the company is using AWS Lambda, with [serverless framework](https://github.com/serverless/serverless), and needless to say, I tried to bring in `esm` 😃 and it works great, although only in bridge mode (I believe AWS Lambda currently doesn't support the `--require` flag as far as I can tell). I noticed something regarding the cache behavior, but I think I better open a new issue for that. If I don't fall asleep I'll also try to open the other issues from some time ago.

you might notice that I'm creating a `js` folder as well as an `mjs` folder. I think I had to do some additional "manual" steps (commenting/uncomenting)  in order to have it running locally, as a run against `node --experimental modules`, as well as against `esm strict mode`.

if you want you can ignore that for now, and I'll add it later on again. I wanted to ask you some time how you wanted the tests to run ideally once merged in.

Let me know if you have any questions.
